onnxruntime/core/providers/mkldnn/subgraph/mkldnn_custom_op.h

```diff
@@ -0,0 +1,283 @@
+      if (status.IsOK())
+		kernel->ReorderWeights(input_tensors, cpu_engine_);
+    }
+  }
+
+  SubgraphContext context_;
+  mkldnn::engine& cpu_engine_;
+};
+
+// Pool which allows for reuse of MKLDNN Conv primitives which are expensive to instantiate.
+// To address thread safety, the primitives are stored in a map on thread local storage.
+template <typename T>
+class SubgraphPrimitivePool : public PrimitivePool<T> {
+ public:
+  static SubgraphPrimitive<T>* Get(const ONNXRunTimeTensor* input_tensors,
+                                   const std::string subgraph_key,
+                                   const SubgraphParams& params) {
+    SubgraphPrimitive<T>* primitive = dynamic_cast<SubgraphPrimitive<T>*>(
+        SubgraphPrimitivePool<T>::GetInstance().GetPrimitive(subgraph_key));
+
+    if (primitive == nullptr) {
+      auto subgraph_primitive = std::make_unique<SubgraphPrimitive<T>>(input_tensors, params);
+      primitive = subgraph_primitive.get();
+      SubgraphPrimitivePool<T>::GetInstance().SetPrimitive(subgraph_key, std::move(subgraph_primitive));
+    }
+    return primitive;
+  }
+
+ private:
+  SubgraphPrimitivePool() = default;
+  ~SubgraphPrimitivePool() = default;
+
+  static SubgraphPrimitivePool& GetInstance() {
+    static SubgraphPrimitivePool pool;
+    return pool;
+  }
+};
+}  // namespace
+
+template <typename T>
+class MkldnnCustomOp {
+ public:
+  explicit MkldnnCustomOp(const ComputeContext* context,
+                          const std::unordered_map<std::string, ONNX_NAMESPACE::AttributeProto>& attributes,
+                          MKLDNNExecutionProvider* provider) {
+    params_.provider = provider;
+    params_.attributes = attributes;
+    params_.mkl_context.reset(new MKLContext(context->allocate_func, context->release_func, context->allocator_handle));
+
+    auto sub_it = attributes.find("subgraph_id");
+    if (sub_it->second.type() != ::ONNX_NAMESPACE::AttributeProto_AttributeType::AttributeProto_AttributeType_STRING) {
+      subgraph_id_ = sub_it->second.s();
+      params_.subgraph = provider->GetMklSubgraph(subgraph_id_);
+      std::ostringstream key_os;
+      key_os << subgraph_id_ << "-" << params_.subgraph->mklnodes.back().name << "-" << params_.subgraph->mklnodes.back().output_name;
+      subgraph_key_ = key_os.str();
+    }
+  }
+
+  Status Compute(const ONNXRunTimeTensor* input_tensors, const size_t num_inputs,
+                 ONNXRunTimeTensor* const output_tensors, const size_t num_outputs) const {
+    Status status;
+    ORT_UNUSED_PARAMETER(num_inputs);
+    ORT_UNUSED_PARAMETER(num_outputs);
+
+    try {
+      auto xshape = input_tensors[0].shape;
+      auto xdim = input_tensors[0].ndim;
+      TensorShape x_shape(xshape, xdim);
+      mkldnn::memory::dims src_dims(x_shape.GetDims().begin(), x_shape.GetDims().end());
+      std::string key;
+      key.reserve(128);
+      key = subgraph_key_;
```

Comment:
Can AddDimsToKey be written such that it accepts a const string and returns the newly formed key? Moreover, we should rename this to GenerateKey() as addition is just one way to generate the key.

Reply:
@pranavsharma  AddDimsToKey is widely used in Vanilla mkldnn operators. Should I go ahead and modify it?